functional-gax/src/main/scala/com/permutive/google/gax/ResponseStream.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2022 Permutive
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.permutive.google.gax
+
+import cats.effect.kernel.{Async, Deferred, Resource, Sync}
+import cats.effect.std.{Dispatcher, Queue}
+import cats.syntax.all._
+import com.google.api.gax.rpc.{ResponseObserver, StreamController}
+import fs2.Stream
+
+//TODO: tests
+private[gax] object ResponseStream {
+
+  def stream[F[_]: Async, A](
+      listen: ResponseObserver[A] => F[Unit],
+      chunkSize: Int
+  ): Stream[F, A] = for {
+    dispatcher <- Stream.resource(Dispatcher.sequential[F])
+    controller <- Stream.eval(Deferred[F, StreamController])
+    queue <- Stream.eval(Queue.bounded[F, Option[A]](chunkSize))
+    err <- Stream.eval(Deferred[F, Throwable])
+    observer = new QueuedResponseObserver[F, A](
+      queue,
+      dispatcher,
+      chunkSize,
+      err,
+      controller
+    )
+    _ <- Stream.bracketCase(listen(observer)) {
+      // Cancellation added here once ResponseObserver#onStart should have been called
+      case (_, Resource.ExitCase.Canceled) => observer.cancel
+      case (_, _) => Sync[F].unit
+    }
+    res <- observer.stream
+  } yield res
+
+  private[ResponseStream] class QueuedResponseObserver[F[_]: Sync, A] private[ResponseStream] (
+      queue: Queue[F, Option[A]],
+      dispatcher: Dispatcher[F],
+      chunkSize: Int,
+      err: Deferred[F, Throwable],
+      controller: Deferred[F, StreamController]
+  ) extends ResponseObserver[A] {
+
+    val stream: Stream[F, A] = for {
+      c <- Stream.eval(controller.get)
+      req = Sync[F].delay(c.request(chunkSize))
+      items <- Stream
+        .evalSeq(req *> queue.take.product(queue.tryTakeN(Some(chunkSize - 1))).map { case (hd, tl) =>
```

Comment:
This should be a slightly better encoding of demand - ensures that the first `take` is blocking (since the `req` won't provide values immediately) and then tries to take the 'rest' of the chunk without blocking